File: tweet2csv.cc
#include <iostream>
#include <regex>
#include <string>

std::regex hashtag("(#+[a-zA-Z0-9(_)]{1,})");
std::regex username("(@+[a-zA-Z0-9(_)]{1,})");

class Tweet {
 public:
  std::string time;
  std::string user;
  std::string tweet;
  std::string hashtags;
  std::string mentions;

  Tweet() : time(""), user(""), tweet(""), hashtags(""), mentions("") {}

  std::string output() {
    hashtags = std::regex_replace(tweet, hashtag, "$1");
    mentions = std::regex_replace(tweet, username, "$1");
    return "\"" + time + "\",\"" + user + "\",\"" + tweet + "\",\"" + mentions + "\",\"" + hashtags + "\"";
  }

  std::string strip(const std::string& s) {
    return s.substr(s.find_first_of('\t') + 1);
  }

  void itime(const std::string& s) {
    time = strip(s);
  }

  void iuser(const std::string& s) {
    user = strip(s).replace("http://twitter.com/", "@");
  }

  void itweet(const std::string& s) {
    tweet = strip(s);
  }
};

int main() {
  Tweet t;
  std::cout << "starting" << std::endl;

  std::string line;
  while (std::getline(std::cin, line)) {
    if (line.empty()) {
      std::cout << t.output() << std::endl;
      t = Tweet();
    }

    if (line.find("T\t") == 0) {
      t.itime(line);
    }

    if (line.find("U\t") == 0) {
      t.iuser(line);
    }

    if (line.find("W\t") == 0) {
      t.itweet(line);
    }
  }

  return 0;
}

File: tweet2csv.py
import sys
import re
from pprint import pprint

dehash = re.compile(r"(#+[a-zA-Z0-9(_)]{1,})")
demention = re.compile(r"(@+[a-zA-Z0-9(_)]{1,})")

class Tweet:

    def __init__(self):
        self.time = ""
        self.user = ""
        self.tweet = ""
        self.hashtag = []
        self.mention = []

    def output(self):
        return (
            f"\"{self.time}\",\"{self.user}\",\"{self.tweet}\",\"{self.mention}\",\"{self.hashtag}\""
            )
    
    def trim(self, s):
        s = s.strip().split("\t")
        if(len(s)>1):
            return(s[1])

    def postcode(self, s):
        return(s.encode("utf-8"))
    
    def postencode(self):
        self.time = self.postcode(self.time)
        self.user = self.postcode(self.user)
        self.tweet = self.postcode(self.tweet)


    def itime(self, s):
        self.time = self.trim(s)

    def iuser(self, s):
        self.user = self.trim(s).replace("http://twitter.com/", "@")

    def itweet(self, s):
        self.tweet = self.trim(s)

    def ihashtag(self):
        s = dehash.findall(self.tweet)
        self.hashtag = s.sort() 

    def imention(self):
        s = demention.findall(self.tweet)
        self.mention = s.sort()
        

t = Tweet()

for l in sys.stdin:

    if '' == l.strip():
        t.imention()
        t.ihashtag()
        t.postencode()
        print(t.output())
        t = Tweet()

    if l.startswith("T\t"):
        t.itime(l)

    if l.startswith("U\t"):
        t.iuser(l)

    if l.startswith("W\t"):
        t.itweet(l)
